### PR TITLE
Move to NuGet batch signing

### PR DESCRIPTION
**Pull request by jaredpar**
This changes our infrastructure to take advantage of the SignTools ability to batch sign NuGet files now. 

Previously we had to build in phases in order to get the proper assets out. Essentially build, sign, package and then build some more. Now though we can simplify this into build, sign, publish. 

This change also cleans up the Insertion directory. I wanted to make that a separate change. However it contains a number of VSIX that embed binaries which must be signed. Hence had to move those VSIX into the build phase in order to have them properly re-packed with signed assets.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @tmat, @jasonmalinowski, @tannergooding, @dotnet/roslyn-infrastructure

File: build/config/SignToolData.json [hunk 1]
```diff
@@ -45,7 +46,9 @@
         "Dlls\\RazorServiceHub\\Microsoft.CodeAnalysis.Remote.Razor.ServiceHub.dll",
         "Dlls\\RazorVisualStudio\\Microsoft.VisualStudio.LanguageServices.Razor.RemoteClient.dll",
         "Dlls\\TextEditorFeatures\\Microsoft.CodeAnalysis.EditorFeatures.Text.dll",
+        "Dlls\\TestUtilities\\net461\\Roslyn.Test.Utilities.dll",
```
Comment (jaredpar) [inline build/config/SignToolData.json]: This may be unnecessary now. Originally it was signed because it was a part of the VS.External nupkg. Discussions on slack caused me to no longer sign that nupkg.

File: build/config/SignToolData.json [hunk 2]
```diff
@@ -100,46 +103,82 @@
...
+      "strongName": null,
+      "values": [
+        "Nuget\\PerBuildPreRelease\\*.nupkg",
+        "Nuget\\PreRelease\\*.nupkg",
+        "Nuget\\Release\\*.nupkg"
       ]
     }
   ],
 
   "exclude": [
```
Comment (jasonmalinowski) [inline build/config/SignToolData.json]: Is there a better way to manage this list? It's getting pretty big and ugly...

File: build/scripts/build.ps1 [hunk 3]
```diff
@@ -458,13 +504,14 @@ function Deploy-VsixViaTool() {
 function Run-SignTool() { 
     Push-Location $repoDir
     try {
+
```
Comment (jasonmalinowski) [inline build/scripts/build.ps1]: Unexpected extra line.

File: build/scripts/build.ps1 [hunk 4]
```diff
@@ -209,6 +210,51 @@ function Build-ExtraSignArtifacts() {
     }
 }
 
+function Build-InsertionItems() {
```
Comment (jasonmalinowski) [inline build/scripts/build.ps1]: Is there a good way to run this without having to build Roslyn.sln each time?

File: src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb [hunk 5]
```diff
@@ -831,6 +827,15 @@ Public Class BuildDevDivInsertionFiles
                     Continue For
                 End If
 
+                If parent.EndsWith("DevDivPackages\Roslyn", StringComparison.OrdinalIgnoreCase) Then
+                    Continue For
+                End If
+
+                ' Wild cards aren't supported
+                If item.Contains("*") Then
+                    Continue For
```
Comment (jasonmalinowski) [inline src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb]: Is this just to prevent somebody from trying to add *? If so, why not throw an exception?

File: src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb [hunk 6]
```diff
@@ -831,6 +827,15 @@ Public Class BuildDevDivInsertionFiles
                     Continue For
                 End If
 
+                If parent.EndsWith("DevDivPackages\Roslyn", StringComparison.OrdinalIgnoreCase) Then
```
Comment (jasonmalinowski) [inline src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb]: Add a comment why?

File: build/config/SignToolData.json (hunk 2 above)
Comment (jaredpar) [reply build/config/SignToolData.json]: Which part: the set of nuget packages or the totality of the file?

File: build/config/SignToolData.json (hunk 1 above)
Comment (jaredpar) [reply build/config/SignToolData.json]: This is necessary. They are both included in our MS.VS.IntegrationTest.Utilities NuPkg

File: build/config/SignToolData.json (hunk 2 above)
Comment (weshaggard) [reply build/config/SignToolData.json]: FWIW for the core repo's we drop a requires_signing file (https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/sign.targets#L86) in the bin folder for any binary that we need to sign. That file contains the information on which strong and and certificate to sign it with. Then as part of our signing step we just find all the requires_signing files and batch sign them. That prevents us form needing to maintain a list of all the files we need to sign.

**Comment by weshaggard** [comment]
>Now though we can simplify this into build, sign, publish.

For the core repos that only produce nuget packages we are already doing a build, sign, publish step feel free to stop by if you want to exchange notes.

**Comment by jaredpar** [comment]
retest ubuntu_16_debug_prtest please

File: build/config/SignToolData.json (hunk 2 above)
Comment (jaredpar) [reply build/config/SignToolData.json]: @weshaggard 

> That prevents us form needing to maintain a list of all the files we need to sign.

We prefer this approach: an explicit manifest of items that need to be signed.

**Comment by jaredpar** [comment]
Signed build now passing. Will merge once the latest changes finish validation.

https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=1106027&_a=summary

File: build/config/SignToolData.json (hunk 2 above)
Comment (jasonmalinowski) [reply build/config/SignToolData.json]: @jaredpar my concern is this seems to be yet another specification of our dependencies, i.e. any time @agocke takes a dependency on a new façade, we also have to add it to this exclusion list. Can we auto-exclude things already signed, or some other heuristic?

Comment (jaredpar) [reply build/config/SignToolData.json]: Gotcha it's the exclusion list you're concerned about. 

> Can we auto-exclude things already signed, or some other heuristic?

The goal of the exclusion list is to help prevent signing errors by fully validating our assets at build time. It's very easy for a new binary to sneak into a VSIX that we didn't intend through innocuous changes to MSBuild files. If that binary is unsigned then suddenly the VSIX is not fully valid anymore. The exclusion list is meant to help with that by identifying the files which are deliberately meant to be there. 

I agree it's clunky though. It's also inexact because it uses file names which is not a strong, or unique, ID. I've considered a couple of solutions: 

1. Allow globs: `System.*`, `Microsoft.Build.*` and we're done ... right? Sounds good until you get into the details. Excluding `System.*` is easy sure but what about `Microsoft.Build.*`? That means we've excluded checking for our own packages. Hence we now need positive and negative excludes. At this point the list is hard to follow and easy for mistakes to happen. Not in favor of this option. 
1. Don't warn if the item is already signed: here we allow binaries which are fully signed to be added to the VSIX / NUPKG without any alteration of the configuration file. This means we still need to list items like NUPKG which can't be signed yet so the exclude list can't go away completely. Overall it's a bit more implicit but may be a good balance.

**Comment by jaredpar** [comment]
https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=1106213&_a=summary

File: build/config/SignToolData.json (hunk 2 above)
Comment (jasonmalinowski) [reply build/config/SignToolData.json]: Would there be some useful feature for "exclude everything that we're consuming from NuGet packages?" Ignoring the details of implementing such a check (you can probably figure it out via groveling either our project.assets.json or the packages directory or both), that means we're more pass-through.

I don't think we can entirely get rid of the exclusion list, but right now it seems having to update it is the norm, not the exception?

Comment (jaredpar) [reply build/config/SignToolData.json]: NuGet content isn't sufficient though. What happens if a downstream repo, or a 3rd party, starts publishing unsigned binaries?

Comment (jasonmalinowski) [reply build/config/SignToolData.json]: Then validate it's signed before excluding it automatically. ;-) I would have presumed this exclusion list just excludes us from signing it, it also validates it's already signed at the same time?

Comment (jaredpar) [reply build/config/SignToolData.json]: Part of the problem here is finding an API that actually tells me if a binary is signed or not: strong and cert. Lacking a good API for that I had to go to an imprecise exclusion list. Given a reliable API for determining signing I would just allow implicit excludes for such binaries.
